File: h_m.js
var request = require('request'),
	cheerio = require('cheerio'),
	db = require('mongoskin').db('weave:weave2013@ds047948.mongolab.com:47948/weave');

var url = "http://www.hm.com/gb/subdepartment/LADIES?Nr=90001#pageSize=MORE&Nr=90001";

function getProducts(url, callback) {
	request(url, function (error, response, body) {
		if (error) throw error;

		$ = cheerio.load(body);
		productList = $("#list-products a");
		// console.log(productList);

		var productLinks = [];

		productList.each(function() {
			productLinks.push($(this).attr("href"));
		})

		//console.log(productLinks);
		callback(productLinks);
	});
};

getProducts(url, function (productLinks) {
	for(link in productLinks) {

		request(productLinks[link], (function(link) {
			return function (error, response, body) {
				if (error) throw error;
				
				$_ = cheerio.load(body);

				var title = $_("h1").text().replace(/\s+/g, ' ').split("£")[0].slice(1);

				var imageUrl = $_("#product-image").attr('src');

				if (imageUrl.substring(0,2) == "//") {
					imageUrl = "http:" + imageUrl;
				};

				var type = $_(".breadcrumbs li:nth-last-child(2)").text().split("/")[0].trim();

				var brand = "H&M";

				var shop = "H&M";

				var price = $_("#text-price span").text()

				var item = {
					"title": title, 
					"url": link,
					"price": price,
					"brand": brand,
					"shop": shop,
					"type": type,
					"imageUrl": imageUrl,
					"collectionDate" : new Date().toDateString()
				};				

				db.collection('products').find({url: link}).toArray(function (err, result) {
					if (err) throw err;
					if (result.length == 0) {
						db.collection('products').insert(item, function (err, result) {
							if (err) throw err;
							console.log("Inserted item " + title)
						});
					}
				});

				//console.log(item);
			}
		})(productLinks[link]));
	}
});

/*var google = "http://www.google.com";
request(google, function (error, response, body) {
  if (!error && response.statusCode == 200) {
    console.log(body) // Print the google web page.
  }
})*/

File: h_m.rb
require 'mechanize'
require 'nokogiri'
require 'open-uri'
require 'mongo'
require 'json'
require './categorize'

agent = Mechanize.new

agent.get("http://www.hm.com/gb/subdepartment/LADIES?Nr=90001")


products = agent.page.search("#list-products a")

product_links = Array.new

products.each do |product|
	product_links << product.attr("href")
end

i = 2

until i == 10 do
	agent.page.link_with(text: "#{i}").click
	products = products = agent.page.search("#list-products a")
	products.each do |product|
		product_links << product.attr("href")
	end
	i += 1
end

product_links

shop = "H&M"

brand = "H&M"

product_links.each do |link|
	agent.get(link)

	title = agent.page.search("h1").text.strip.split("£")[0].split("\r")[0]

	# puts link
	# puts title

	price = "£" + agent.page.search("h1").text.strip.split("£")[1].to_s

	images = agent.page.search("#product-thumbs img")

	imageLinks = Array.new

	images.each do |image|
		large_image = image.attr("src").gsub "thumb", "large"
		imageLinks << "http:" + large_image
	end

	cat = agent.page.search(".breadcrumbs li:nth-last-child(2)").text.strip.split("\r")[0].rstrip

	begin
		# Method categorize must have a category.
		category, subCategory = categorize title, cat
		if subCategory == nil && category == nil
			puts "ALERT =================================="
		end
	rescue Exception => e 
		puts e
	end

	description = agent.page.search(".description p").text.strip.split("\r")[0]

	material = agent.page.search("span#text-information").text + " " + agent.page.search("span#text-careInstruction").text

	item = {
		"title" => title,
		"url" => link,
		"images" => imageLinks,
		"price" => price,
		"shop" => shop,
		"brand" => brand,
		"category" => category,
		"subCategory" => subCategory,
		"materials" => material,
		"description" => description,
		"collectionDate" => Time.now.strftime("%a %b %d %Y")
	}

	# puts item

	uri = "mongodb://andy:weave2013@paulo.mongohq.com:10000/weave-dev"

	client = Mongo::MongoClient.from_uri(uri)

	db_name = uri[%r{/([^/\?]+)(\?|$)}, 1]
	db = client.db(db_name)

	products = db.collection("products")

	# Note that the insert method can take either an array or a single dict.
	if products.find("url" => link).to_a.empty?
		if category != nil
		# 	puts "Title: " + item["title"]
		# 	puts "URL: " + item["url"]
		# 	puts "Image Links: " + item["images"].to_s
		# 	puts "Shop: " + item["shop"]
		# 	puts "Brand: " + item["brand"]
		# 	puts "Category: " + item["category"]
		# 	puts "Subcategory: " + item["subCategory"].to_s
		# 	puts "Materials: " + item["materials"]
		# 	puts "Collection Date: " + item["collectionDate"]
		# 	print "Insert into database? (y/n):"
		# 	confirm = gets
		# 	confirm.chomp!
		# 	if confirm == "y" or confirm == "Y"
		# 		products.insert(item)
		# 		puts "Item inserted"
		# 	else
		# 		puts "You decided not to insert into the database, Also Nick is a dick."
		# 	end

			products.insert(item)
			puts title + " inserted"
		else
			puts "Did not insert " + title + " because the category or sub category are undefined. Check " + link
		end
	else
		puts title + " already exists in the database"
	end

	# puts title + ". Category: " + category.to_s + ". Subcategory: " + subCategory.to_s


	# puts imageUrls
	# puts title + " " + price
end

# £

	